plugins/processors/ec2tagger/ec2tagger.go

```diff
@@ -418,44 +366,47 @@ func (t *Tagger) refreshLoopToUpdateTagsAndVolumes() {
 	}
 }
 
-// updateVolumes calls EC2 describe volume
-func (t *Tagger) updateVolumes() error {
-	if t.ebsVolume == nil {
-		t.ebsVolume = NewEbsVolume()
+/*
+	Retrieve metadata from IMDSv2,IMDSv1 and use these metadata to:
+	* Extract InstanceID, ImageID, InstanceType to create custom dimension for collected metrics
+	* Extract InstanceID to retrieve Instance's Volume and Tags
+	* Extract Region to create aws session with custom configuration
+	For more information on IMDSv2, please follow this document https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+*/
+func (t *Tagger) initialRetrievalOfEC2Metadata() error {
+	for _, tag := range t.EC2MetadataTags {
+		switch tag {
+		case mdKeyInstanceId:
+			t.ec2MetadataLookup.instanceId = true
+		case mdKeyImageId:
+			t.ec2MetadataLookup.imageId = true
+		case mdKeyInstanceType:
+			t.ec2MetadataLookup.instanceType = true
+		default:
+			t.Log.Errorf("ec2tagger: Unsupported EC2 Metadata key: %s.", tag)
+		}
 	}
 
-	input := &ec2.DescribeVolumesInput{
-		Filters: []*ec2.Filter{
-			{
-				Name:   aws.String("attachment.instance-id"),
-				Values: aws.StringSlice([]string{t.instanceId}),
-			},
-		},
+	t.Log.Infof("ec2tagger: Check EC2 Metadata.")
+	doc, err := t.ec2MetadataProvider().GetInstanceIdentityDocument()
+	if err != nil {
+		t.Log.Error("ec2tagger: Timeout when retrieving EC2 Metadata. This plugin must only be used on an EC2 instance.")
+		if context.CurrentContext().RunInContainer() {
+			t.Log.Warn("ec2tagger: Timeout may occurred because hop limit is too small. Please increase hop limit to 2 by following this document https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-options.html#configuring-IMDS-existing-instances.")
+		}
+		return err
 	}
 
-	for {
-		result, err := t.ec2.DescribeVolumes(input)
-		if err != nil {
-			return err
-		}
-		for _, volume := range result.Volumes {
-			for _, attachment := range volume.Attachments {
-				t.ebsVolume.addEbsVolumeMapping(volume.AvailabilityZone, attachment)
-			}
-		}
-		if result.NextToken == nil {
-			break
-		}
-		input.SetNextToken(*result.NextToken)
+	t.ec2MetadataRespond.region = doc.Region
+	t.ec2MetadataRespond.instanceId = doc.InstanceID
```

Comment:
Why do we not check the lookup for instance ID like the other values below?

Reply:
InstanceID is a required field, instead of **optional field** since InstanceID can be used to retrieve EC2 Instance Volumes and Tags (e.g [this line](https://github.com/aws/amazon-cloudwatch-agent/pull/480/files#diff-6638b992018408bf44cac12361171fd050ff44fcc24c28f326988f4768a6dc2dR102). Therefore, it is not optional to check the  lookup for instanceID